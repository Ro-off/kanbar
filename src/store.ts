import { configureStore } from "@reduxjs/toolkit";
import createSagaMiddleware from "redux-saga";

import { tasksWatcher } from "./sagas/tasksSaga";
import { tasksReducer } from "./reducers/tasksReducer";

const sagaMiddleware = createSagaMiddleware();

export const store = configureStore({
  reducer: tasksReducer,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(sagaMiddleware),
});

sagaMiddleware.run(tasksWatcher);
