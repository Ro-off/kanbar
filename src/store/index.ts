export * from "./configureStore";
export * from "./tasks";
