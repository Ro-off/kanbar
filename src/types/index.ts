import { SVGProps } from "react";

import { ColumnIds } from "@/components/taskRow";

export type IconSvgProps = SVGProps<SVGSVGElement> & {
  size?: number;
};

export interface Task {
  id: string;
  description?: string;
  newField?: boolean;
  columnId: ColumnIds;

  [key: string]: any;
}

export interface AddTaskAction {
  type: "ADD_TASK";
  payload: ColumnIds;
}

export interface RemoveTaskAction {
  type: "REMOVE_TASK";
  payload: string;
}

export interface UpdateTaskAction {
  type: "UPDATE_TASK";
  payload: {
    id: string;
    data: Partial<Task>;
    index?: number;
  };
}

export type TaskAction = AddTaskAction | RemoveTaskAction | UpdateTaskAction;

export type TasksState = {
  tasks: Task[];
};

export interface TaskCardProps extends Task {
  isEditing?: boolean;
  index?: number;
}
