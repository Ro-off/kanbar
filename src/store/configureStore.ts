import { configureStore } from "@reduxjs/toolkit";
import createSagaMiddleware from "redux-saga";

import { tasksWatcher } from "./tasks";
import tasksReducer from "./tasks";

const sagaMiddleware = createSagaMiddleware();

export const store = configureStore({
  reducer: tasksReducer,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(sagaMiddleware),
});

sagaMiddleware.run(tasksWatcher);
