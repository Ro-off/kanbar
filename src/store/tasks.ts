import { takeEvery } from "redux-saga/effects";
import { ColumnIds } from "@types";

import {
  AddTaskAction,
  RemoveTaskAction,
  Task,
  TaskAction,
  TasksState,
  UpdateTaskAction,
} from "../types";

import { store } from "./configureStore";

const ADD_TASK = "ADD_TASK";
const REMOVE_TASK = "REMOVE_TASK";
const UPDATE_TASK = "UPDATE_TASK";

const delay = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));

export const updateTask = (id: string, updatedFields: Partial<Task>) => {
  store.dispatch({
    type: UPDATE_TASK,
    payload: {
      id,
      ...updatedFields,
    },
  });
};

export const deleteTask = (id: string) => {
  store.dispatch({
    type: REMOVE_TASK,
    payload: id,
  });
};

export const createTask = (columnId: ColumnIds) =>
  store.dispatch({
    type: ADD_TASK,
    payload: columnId,
  });

export const updateTaskOnBack = async (
  action: UpdateTaskAction,
): Promise<Task[] | undefined> => {
  await delay(1000);

  const tasksStr = localStorage.getItem("tasks");

  if (!tasksStr) {
    localStorage.setItem(
      "tasks",
      JSON.stringify([{ ...action.payload, newField: false }]),
    );

    return;
  }
  const oldTasks: Task[] = JSON.parse(tasksStr);
  const oldIndex = oldTasks.findIndex((task) => task.id === action.payload.id);

  if (oldIndex === -1) {
    const newTasks = [{ ...action.payload, newField: false }, ...oldTasks];

    localStorage.setItem("tasks", JSON.stringify(newTasks));

    return newTasks;
  }

  const oldTask = oldTasks[oldIndex];
  const newTask = { ...oldTask, ...action.payload, newField: false };
  const newIndex =
    action.payload.index !== undefined ? action.payload.index : oldIndex;

  const tasksWithoutOld = [
    ...oldTasks.slice(0, oldIndex),
    ...oldTasks.slice(oldIndex + 1),
  ];

  const newTasks = [
    ...tasksWithoutOld.slice(0, newIndex),
    newTask,
    ...tasksWithoutOld.slice(newIndex),
  ];

  localStorage.setItem("tasks", JSON.stringify(newTasks));

  return newTasks;
};

export const getTasksOnBack = async (): Promise<Task[] | undefined> => {
  await delay(1000);

  const tasksStr = localStorage.getItem("tasks");

  if (!tasksStr) {
    return undefined;
  }

  return JSON.parse(tasksStr);
};

export const removeTaskOnBack = async (action: RemoveTaskAction) => {
  await delay(1000);

  const tasksStr = localStorage.getItem("tasks");

  if (!tasksStr) {
    return;
  }

  const oldTasks: Task[] = JSON.parse(tasksStr);
  const oldIndex = oldTasks.findIndex((task) => task.id === action.payload);

  if (oldIndex === -1) {
    return;
  }

  const tasksWithoutOld = [
    ...oldTasks.slice(0, oldIndex),
    ...oldTasks.slice(oldIndex + 1),
  ];

  localStorage.setItem("tasks", JSON.stringify(tasksWithoutOld));
};

function* updateTaskWorker(action: UpdateTaskAction) {
  yield delay(1000);
  updateTaskOnBack(action);
}

function* removeTaskWorker(action: RemoveTaskAction) {
  yield delay(1000);
  removeTaskOnBack(action);
}

export function* tasksWatcher() {
  yield takeEvery("REMOVE_TASK", removeTaskWorker);
  yield takeEvery("UPDATE_TASK", updateTaskWorker);
}

let initialState: TasksState = {
  tasks: (await getTasksOnBack()) ?? [],
};

const tasksReducer = (
  state: TasksState = initialState,
  action: TaskAction,
): TasksState => {
  switch (action.type) {
    case ADD_TASK:
      return {
        tasks: [
          {
            id: self.crypto.randomUUID(),
            description: "",
            columnId: action.payload as AddTaskAction["payload"],
            newField: true,
          },
          ...state.tasks,
        ],
      };
    case REMOVE_TASK:
      return {
        ...state,
        tasks: state.tasks.filter((task) => task.id !== action.payload),
      };

    case UPDATE_TASK:
      let taskArray = [...state.tasks];
      const payload = action.payload as UpdateTaskAction["payload"];
      const oldTask = taskArray.find((task) => task.id === payload.id);

      if (!oldTask) {
        return state;
      }
      const newTask = { ...oldTask, ...payload, newField: false };

      const oldIndex = taskArray.indexOf(oldTask);
      const newIndex = payload.index !== undefined ? payload.index : oldIndex;

      taskArray.splice(oldIndex, 1);

      return {
        ...state,
        tasks: [
          ...taskArray.slice(0, newIndex),
          newTask,
          ...taskArray.slice(newIndex, taskArray.length),
        ],
      };

    default:
      return state;
  }
};

export default tasksReducer;
