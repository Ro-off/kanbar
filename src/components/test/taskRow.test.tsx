import { expect, it } from "vitest";
import { render } from "@testing-library/react";
import { TaskRow } from "@components";
import { Provider } from "react-redux";

import { TaskRowProps } from "@/types";
import { store } from "@/store";

const taskRowTestProps: TaskRowProps = {
  columnId: "todo",
};

it("Check taskRow snapshot", () => {
  const { asFragment } = render(
    <Provider store={store}>
      <TaskRow {...taskRowTestProps} />
    </Provider>,
  );

  expect(asFragment()).toMatchSnapshot();
});
