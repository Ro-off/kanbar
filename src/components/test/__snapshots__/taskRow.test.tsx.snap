// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Check taskRow snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col relative overflow-hidden text-foreground box-border bg-content1 outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 shadow-medium rounded-large transition-transform-background motion-reduce:transition-none xl:w-96 w-72 h-full"
    tabindex="-1"
  >
    <div
      class="z-10 w-full items-center shrink-0 overflow-inherit color-inherit subpixel-antialiased flex flex-row justify-between opacity-80 bg-blue-100 p-4 rounded-t-lg"
    >
      <h2
        class="text-lg font-semibold"
      >
        To Do
      </h2>
      <button
        class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal subpixel-antialiased overflow-hidden tap-highlight-transparent transform-gpu data-[pressed=true]:scale-[0.97] outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 text-small gap-2 rounded-full px-0 !gap-0 transition-transform-colors-opacity motion-reduce:transition-none bg-transparent text-default-foreground data-[hover=true]:bg-default/40 min-w-10 w-10 h-10 mt-2"
        data-react-aria-pressable="true"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="h-5 w-5"
          fill="none"
          focusable="false"
          height="1em"
          role="presentation"
          viewBox="0 0 24 24"
          width="1em"
        >
          <path
            d="M12 5V19"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
          <path
            d="M5 12H19"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
        <span
          class="sr-only"
        >
          Add Task
        </span>
      </button>
    </div>
    <hr
      class="shrink-0 bg-divider border-none w-full h-divider"
      role="separator"
    />
    <div
      class="relative flex flex-auto flex-col place-content-inherit align-items-inherit h-auto break-words text-left overflow-y-auto subpixel-antialiased p-0 w-full"
    >
      <div
        class="overflow-y-auto data-[top-scroll=true]:[mask-image:linear-gradient(0deg,#000_calc(100%_-_var(--scroll-shadow-size)),transparent)] data-[bottom-scroll=true]:[mask-image:linear-gradient(180deg,#000_calc(100%_-_var(--scroll-shadow-size)),transparent)] data-[top-bottom-scroll=true]:[mask-image:linear-gradient(#000,#000,transparent_0,#000_var(--scroll-shadow-size),#000_calc(100%_-_var(--scroll-shadow-size)),transparent)] h-full"
        data-bottom-scroll="false"
        data-orientation="vertical"
        data-top-scroll="false"
        style="--scroll-shadow-size: 40px;"
      >
        <div
          class="w-full overflow-x-hidden h-full"
        >
          <p
            class="text-gray-500 text-center m-4"
          >
            No tasks available.
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
